### PR TITLE
SAMZA-2579: Force restart feature for Container Placements

### PR DESCRIPTION
**Pull request by Sanil15**
**Changes**: The current restart ability for container placements works in the following way:
1. Tries to fetch resources on a host
2. Stops the active container if resources are accrued
3. Tried to start the container on host accrued

In production, we have seen the following observation at Linkedin
1. Some jobs are configured to use resources for the peak which leads to no headroom left on a host for requesting additional resources
2. This leads to restart requests failing due to not able to get resources on that host

A fix to this is to implement a force-restart utility , in this version we will stop the container first and then accrue resources. The upside being we will at least free up the resources on the host before issuing resource request, the downside being it will be a best-effort scenario to bring that container back up on that host

**API Changes:** Added new param values to destinationHost param for container placement request message

LAST_SEEN: Tries to restart a container on last seen host with RESERVE -> STOP -> MOVE policy

FORCE_RESTART_LAST_SEEN: Tries to restart a container on last seen host with STOP -> RESERVE -> MOVE policy

**Tests:** Tested the change with a yarn job

**Upgrade Instructions:** None

**Usage Instructions:** None

### CODE REVIEW

**Comment by Sanil15** [comment]
@rmatharu please have a look

File: samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java [hunk 1]
```diff
@@ -54,6 +54,8 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(ContainerManager.class);
   private static final String ANY_HOST = ResourceRequestState.ANY_HOST;
+  private static final String LAST_SEEN = "LAST_SEEN";
+  private static final String FORCE_RESTART_LAST_SEEN = "FORCE_RESTART_LAST_SEEN";
```
Comment (rmatharu) [inline samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java:58]: Isnt force-restart a misnomer here?
All its really doing is issuing a stop and relying on the AM to bring it back up?

In a corner case, AM will not be able to bring it back up on the same host, rather on a different available host?

File: samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java [hunk 2]
```diff
@@ -391,6 +402,28 @@ public void registerContainerPlacementAction(ContainerPlacementRequestMessage re
...
+      clusterResourceManager.stopStreamProcessor(samzaApplicationState.runningProcessors.get(processorId));
+      writeContainerPlacementResponseMessage(requestMessage, ContainerPlacementMessage.StatusCode.SUCCEEDED,
+          "Successfully issued a stop container request falling back to normal restart path");
+      return;
+    }
+
+    /**
+     * When destination host is {@code LAST_SEEN} its treated as a restart request on the host where container is running
+     * on or has been seen last
+     */
```
Comment (rmatharu) [inline samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java:420]: This comment has the same meaning as the one above 
"if (destinationHost.equals(FORCE_RESTART_LAST_SEEN))" 

Could we write the difference

File: samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java (hunk 1 above)
Comment (Sanil15) [reply samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java:58]: discussed offline, its the destinationHost param which is.

**All its really doing is issuing a stop and relying on the AM to bring it back up?**
Yup that is what is it doing

**In a corner case, AM will not be able to bring it back up on the same host, rather on a different available host?**
Yes and that is documented

File: samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java (hunk 2 above)
Comment (Sanil15) [reply samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java:420]: sure

**Comment by Sanil15** [comment]
> need to call it kill-container maybe

actual tool will be sometime like: `samza-tool kill-container ....`

File: samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java [hunk 3]
```diff
@@ -391,6 +402,30 @@ public void registerContainerPlacementAction(ContainerPlacementRequestMessage re
       return;
     }
 
+    /*
+     * When destination host is {@code FORCE_RESTART_LAST_SEEN} its treated as eqvivalent to kill -9 operation for the container
```
Comment (rmatharu) [inline samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java:406]: small thing: 
I think its equivalent to a "kill"  (i.e., sigstop), not "kill -9" (sigkill). 
Could you verify using the yarn docs

**Pull request merged**